sunkit_image/nafe.py

```diff
@@ -0,0 +1,411 @@
+                              sigma=nrs, weight=nafe_weight, imargin=imargin,
+                              n=n, c=c)
+
+        # Evaluate indices
+        nrows, ncols = map_data.shape
+        rows, cols = np.meshgrid(range(nrows), range(ncols), indexing='ij')
+        indices = zip(rows.flat, cols.flat)
+
+        # NAFE multiprocessing context
+        with Pool(nproc) as pool:
+            # return margin [0, 2 ** 16 - 1]
+            nafe_image = list(tqdm.tqdm(pool.imap(nafe_instance.transform, indices), total=map_data.size))
+
+        enhanced_image = np.array(nafe_image).reshape(nrows, ncols)
+        enhanced_image = _transform(old_margin=(0., 2. ** 16 - 1.), new_margin=omargin, old_value=enhanced_image)
+
+    return enhanced_image
+
+
+def _get_membership_grade(n=129, wl=[1 for _ in range(12)]):
+    """
+    Calculates the matrix used for expressing the membership grade of
+    neighborhood pixels to the fuzzy histogram computed for each pixel
+    in image. The matrix L is a linear combination of M Gaussian kernels
+    calculated as:
+
+        L(x, y) = d * sum[m=1:M](wl[m] * G(sigma[m], x, y))
+
+    where:
+        x = horizontal distance from center
+        y = vertical distance from center
+        sigma[m] = 2 ** (m / 2)
+        M = len(wl)
+
+        d = 2 * pi / sum[m=1:M]( wl[m] * (sigma[m] ** - (1/2)) )
+        G(s, x, y) = exp(−0.5 * (x ** 2) + (y ** 2)) * (s ** -2)) / (2 * pi * sqrt(s))    , s = sigma[m]
+
+    Parameters
+    ----------
+    n: : `~int`
+        Neighborhood width. Must be an odd integer and defaults to 129.
+
+    wl : `~list`
+        Weights list of for each Gaussian kernel of the M Gaussian kernels
+        used for computing L. M is calculated as len(wl). Defaults to a
+        1x12 matrix of ones.
+
+    Returns
+    -------
+    L : `~numpy.ndarray`
+        Membership grade matrix.
+
+    Example
+    -------
+    >>> _get_membership_grade(5)
+    array([[0.65289406, 0.73715583, 0.7741782 , 0.73715583, 0.65289406],
+           [0.73715583, 0.86813217, 0.92826105, 0.86813217, 0.73715583],
+           [0.7741782 , 0.92826105, 1.        , 0.92826105, 0.7741782 ],
+           [0.73715583, 0.86813217, 0.92826105, 0.86813217, 0.73715583],
+           [0.65289406, 0.73715583, 0.7741782 , 0.73715583, 0.65289406]])
+    """
+    M = len(wl)
+    sigma = [2 ** (m / 2) for m in range(1, M+1)]  # starts from 1 for maths
+    d = 2 * np.pi / np.sum(wl * np.power(sigma, -1/2))
+
+    wlG = 0
+    for s, w in zip(sigma, wl):
+        G = lambda r, c: np.exp(-0.5 * ((r - n//2) ** 2 + (c - n//2) ** 2) * (s ** -2)) / (2 * np.pi * np.sqrt(s))
+        wlG += w * np.fromfunction(G, (n, n))
+
+    return d * wlG
+
+
+def _transform(old_margin, new_margin, old_value, power=1.0):
+    """
+    # TODO: add doc
+    """
+    return new_margin[0] + (new_margin[1] - new_margin[0]) * (((old_value - old_margin[0]) / (old_margin[1] - old_margin[0])) ** (1.0 / power))
+
+
+class _Nafe():
```

Comment:
I am not sure I am in favour of having a  class here, I would rather it be a series of functions with the "main" function that will what the user calls.

Reply:
I am curious why would a class not be preferred. I come from a C++ background so that's why it seemed the way to go for me.

Reply:
I guess I don't understand why we need a class here.

Reply:
Basically used for keeping reference of the shared variables across the processes pool. Just making the code more readable and the mapping function cleaner.

Reply:
Is there no other way?

Reply:
Pretty sure yes but that one liner mapping would be a lot messier.

Reply:
This: `pool.imap(nafe_instance.transform, indices), total=map_data.size)` is what I mean. I need to use imap because the order of the results matter since they are pixels values. AFAIK the only other way to pass the arguments to imap would be creating an iterable whose elements will mainly be in the form (row, col, some constant arguments). I thought since they are constant arguments this seemed a typical case for using a class.

Reply:
Well if it's the best way to do it, it can stay.

Reply:
I'm ok with the class, but why not having it all as one? I mean to include the `nafe` function as to execute it?

Reply:
I have no preference but current implementation doesn't require a class anymore.